### PR TITLE
Boxes reports functioning

### LINKED ISSUE #1820: LOD visualization

### PR DESCRIPTION
**Pull request by leandroradusky**
Closes #1820, #1821, #1822 & #1555.

Reporting functioning for all the box purposes, also exporting to PDF and SVG working. I know that we talked on the contrary. Still, I'm submitting one big PR because:
- it was started from an old version of `main` and the rebasing plus applying the changes dividing this PR is much more error-prone; 
- the intermediate PRs had code that changes at the end and revising that code will double the work without the necessity
- the first of the PRs to be submitted before was not much smaller than this one. 

Said this... There are three plots & one visualization, which have some differences from the [mockups](https://projects.invisionapp.com/d/main#/console/4103589/470820372/preview) that will be remarked. 

- **Confusion matrix**, which is displayed on all the visualizations

image

- **Measured signal**, looks like the mockup for `Variants` & `LOD` and in the case of `Challenge` have two differences: distractor samples are grouped in blue bars and virus ones are kept in red, and there is a line to indicate the `threshold`.

For LOD:

image

For Challenge: 

image

- **Limit of detection** which is only displayed in `LOD` purpose and looks different attending client request: each sample has a concentration and a measured signal and represents a point in the plot, the regression line computed for all these points is represented with a dashed line. 

image

- **ROC curve** which is only displayed for `Challenge` purpose. The curve itself is displayed in red, and the TPR and FPR of the selected threshold are updated to accommodate the dashed lines.

image

The `Challenge` report allows the user to update the threshold via a slider: 

image

This threshold will update the confusion matrix, dashed lines mentioned above, and the information card on the left:

chrome-capture-2023-0-11

When the threshold bar is moved `update_threshold` is `GET` from the controller to update variables. For `Variants` and `LOD` threshold is always zero: all the samples contain virus, then they are only negative those containing zero concentration (no virus). 

Also, there are enabled buttons to download SVG and PDF files are functioning:

- For the PDF exporting, a new view was generated, avoiding layout and generating a view that will be "printed" to a file. This file is generated using [this library](https://github.com/eKoopmans/html2pdf.js#progress), which helps with margins/page-orientation/page-breaks, etc. The library wasn't installed with NPM in the project, is dynamically loaded in javascript. Should it be installed?
- For the SVG exporting: after intensive stack overflowing the issue, the only solution working properly was to:

1. generate an <a> element that downloads a generated Blob; along with
2. a function that recursively takes the CSS styles of the elements and writes them into the styles property. This function is called applyInLine. The natural option for this when looking to solve this issue was to use getComputedStyle, but it didn't work properly for the react-generated SVG, while this one, more "sophisticated" did.

Screenshots of the PDF (for `Challenge` purpose, illustrative of the others):

1/3
image

2/3
image

3/3
image

**Important**: this PR rely in the existence of a `distractor` attribute in samples, which was faked during the development and not committed, and it should be merged into `main` after the PRs containing the addition of this attribute are merged

### CODE REVIEW

**Review by ysbaddaden — COMMENTED** [review]
@leandroradusky This is only a partial review, so you can continue working! I'm still pondering about the PDF rendering.

File: app/assets/javascripts/components/box_reports_bar_chart.js.jsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+      margin: {top: 20, right: 20, bottom: 50, left: 50},
+      x_labels: []
+    }
+  },
+
+  componentDidMount: function() {
+    if (!this.props.width) {
+      this.setProps({
+        width: this.props.is_for_report ? 400 : 700,
+        height: this.props.is_for_report ? 300 : 400
```
[the remark this answers is not in the document]
Comment (leandroradusky) [reply app/assets/javascripts/components/box_reports_bar_chart.js.jsx:13]: This was dead code, gone :)

File: app/controllers/samples_reports_controller.rb [hunk 2]
```diff
@@ -77,6 +86,19 @@ def delete
...
+    @threshold = params[:threshold].to_f
+    @minthreshold = params[:minthreshold].to_f
+    @maxthreshold = params[:maxthreshold].to_f
+    @reports_data = measured_signal_data(@samples_report)
+    @samples_without_results_count = @samples_report.samples.where("core_fields NOT LIKE '%measured_signal%'").count
+    @purpose = @samples_report.samples[0].box.purpose
+    return unless authorize_resource(@samples_report, READ_SAMPLES_REPORT)
+
+    render '_pdf_report', :layout => false
+  end
```
[the remark this answers is not in the document]
Comment (leandroradusky) [reply app/controllers/samples_reports_controller.rb:101]: The execution goes well with this logic, but I cannot avoid the browser attempting to display a PDF and failing, so I made only one method, but I use a parameter to render normally of `_pdf_report`.

File: app/helpers/samples_reports_helper.rb [hunk 3]
```diff
@@ -1,14 +1,59 @@
 module SamplesReportsHelper
-  def confusion_matrix_svg(id)
-    "No image"
+  def get_rates(samples, signal)
+    confusion_matrix = Hash.new{0}
```
[the remark this answers is not in the document]
Comment (leandroradusky) [reply app/helpers/samples_reports_helper.rb:3]: I've just called the function from the helper declaring it as a `:helper_method`

File: app/controllers/samples_reports_controller.rb [hunk 4]
```diff
@@ -131,4 +159,58 @@ def boxes_data(boxes)
...
+          else
+            confusion_matrix[:predicted_negative] += 1
+            confusion_matrix[:false_negative] += 1
+          end
+        end
+      end
+    end
+
+    confusion_matrix
+  end
```
[the remark this answers is not in the document]
Comment (leandroradusky) [reply app/controllers/samples_reports_controller.rb:213]: I've done it in the most naive way to make it declarative but it is true that finally is quite complex... I changed it to 

```
def confusion_matrix(samples, threshold)
    confusion_matrix = Hash.new{0}
    
    samples.each do |s|
      next unless s.measured_signal
      if s.concentration == 0 || s.distractor
        s.measured_signal > threshold ? confusion_matrix[:false_positive] += 1 : confusion_matrix[:true_negative] += 1
      else
        s.measured_signal > threshold ? confusion_matrix[:true_positive] += 1 : confusion_matrix[:false_negative] += 1
      end
    end

    confusion_matrix[:actual_positive] = confusion_matrix[:true_positive] + confusion_matrix[:false_negative]
    confusion_matrix[:actual_negative] = confusion_matrix[:true_negative] + confusion_matrix[:false_positive]
    confusion_matrix[:predicted_positive] = confusion_matrix[:true_positive] + confusion_matrix[:false_positive]
    confusion_matrix[:predicted_negative] = confusion_matrix[:true_negative] + confusion_matrix[:false_negative]
    confusion_matrix[:total] = confusion_matrix[:actual_positive] + confusion_matrix[:actual_negative]

    confusion_matrix
  end
```

File: app/views/samples_reports/_confusion_matrix.haml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    .confusion-matrix-subtitle.muted
+      TOTAL
+
+  .confusion-matrix-row
+    .confusion-matrix-subtitle.rotated.muted
+      ACTUAL 
+      %br
+      NEGATIVE
+    .confusion-matrix-cell.gray
+      .number{id: "true_negative"}
```
[the remark this answers is not in the document]
Comment (leandroradusky) [reply app/views/samples_reports/_confusion_matrix.haml:22]: It works :)

**Review by ysbaddaden — APPROVED** [review]
I think it's good as-is.

I still think we should generate the PDF on the backend, but we should work on that in a follow-up PR.

**Pull request merged**